presto-hive/src/main/java/io/prestosql/plugin/hive/aws/AwsCurrentRegionHolder.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.plugin.hive.aws;
+
+import com.amazonaws.regions.Region;
+import com.amazonaws.regions.Regions;
+
+import javax.annotation.Nullable;
+
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * Caches the result of calling {@link Regions#getCurrentRegion()} since accessing EC2 instance
+ * metadata repeatedly can result in being throttled and prevent other metadata accessing operations
+ * such as refreshing instance credentials from working normally
+ */
+public final class AwsCurrentRegionHolder
```

Comment:
Could we do the same thing using a Guava memoized supplier?

Reply:
Looks like `MemoizingSupplier` caches `null` values if they're delegate returns them. However, `Regions.getCurrentRegion()` can return null if you've being rate limited or the metadata API is temporarily unreachable, etc. Always rechecking when the value is null is desirable here.

Reply:
We could have the supplier check for null and throw an exception so that it’s not cached. From the PR description, it sounds like null should be treated as an error anyway, since it will cause downstream failures. Better to throw up front with clear error message.

Reply:
I agree that memoizing supplier should be sufficient.

Reply:
I suppose that's fair, but current callers all null-check the result at the moment and I don't think it's unreasonable to want to attempt to get the current region without knowing for sure that you're running in EC2. Lastly, the `Regions.getCurrentRegion()` implementation has already discarded (after logging) any useful message about *why* the result is null by this point. Anyway, you still feel strongly then we can make it throw and switch to memoizing supplier.

Reply:
Indeed, this needs to work outside of AWS (eg, but not limited to, when running locally).
However, all the usages are `if (pinS3ClientToCurrentRegion)` (for a reason so), and this flag can't be supported outside of AWS, right? (assuming it doesn't take "current region" from whatever is configured in `~/.aws/config`)

Reply:
That’s what I’m afraid of interacting with. Other sources of region detection come into play if a region isn’t explicitly set, and while you could argue that those can violate the intention of “pinning to current region”, I chose to just preserve current behavior with an added cache and mutex.

Reply:
Given the intention to explicitly pin the current region, I'm not sure why the callers check for null (probably the first place did it and the others were copied). But looking at it now, this seems like a bug, and we should fail if the region can't be obtained.

I also notice that all the calls are on startup, and are against the instance local metadata API, so I'm curious why this call would happen frequently enough to involve rate limiting?

Reply:
First point about region determination:
There's "current region" as in `Regions.getCurrentRegion()` aka: EC2 instance metadata derived region. That's *probably* the right interpretation for "pinning", but region can also be determined by setting `AWS_DEFAULT_REGION` in an environment variable or using system properties or `~/.aws/config` and probably more I can't think of right now. Leaving the region unset in the client builder will attempt to determine a region from those other sources. It's possible people are relying on that behavior so if we want to change the semantics there I suggest we do it in another PR.

About how often this is actually called:
It's called quite a bit more frequently than that in our particular setup, but in a "typical" deployment you're right- these won't be all that frequent. If dynamically creating / removing catalogs gets better support, doing that could induce failures.